README.md
<br />
<div align="center">
  <a href="https://www.npmjs.com/package/mjml-loader">
    <img src="https://img.shields.io/npm/v/mjml-loader.svg?maxAge=86400" alt="Last npm Registry Version">
  </a>
  <a href="https://travis-ci.org/ramasilveyra/mjml-loader?branch=master">
    <img src="https://travis-ci.org/ramasilveyra/mjml-loader.svg?branch=master" alt="Build Status">
  </a>
  <a href="https://codecov.io/github/ramasilveyra/mjml-loader?branch=master">
    <img src="https://img.shields.io/codecov/c/github/ramasilveyra/mjml-loader.svg?branch=master" alt="Code coverage">
  </a>
</div>

<h1 align="center">MJML loader for webpack</h1>

<p align="center">Converts MJML files to HTML and exports them as string.</p>

<h2 align="center">Table of Contents</h2>

- [Install](#install)
- [Usage](#usage)
- [Examples](#examples)
- [Contribute](#contribute)
- [License](#license)

<h2 align="center">Install</h2>

**Node.js v6.5 or newer** is required.

**webpack 2 or newer** is required.

Via the yarn client:

```bash
$ yarn add --dev mjml-loader webpack
```

Via the npm client:

```bash
$ npm install --save-dev mjml-loader webpack
```

<h2 align="center">Usage</h2>


[Look into the official webpack loaders usage documentation](https://webpack.js.org/concepts/loaders/#using-loaders).

At this moment this loader has no options.

<h2 align="center">Examples</h2>

### Compile MJML file to HTML file

For transforming a MJML file to a HTML file is better to use [`mjml cli`](https://www.npmjs.com/package/mjml-cli) or [`gulp-mjml`](https://www.npmjs.com/package/gulp-mjml).

### Inline the MJML result (HTML) into bundle

With this configuration:

```javascript
{
  // ...
  module: {
    rules: [
      { test: /\.mjml$/, loader: 'mjml-loader' }
    ]
  }
  // ...
}
```

With this MJML file:

`template.mjml`
```xml
 <mjml>
  <mj-body>
    <mj-container>
      <mj-section>
        <mj-column>
          <mj-text>Hello World!</mj-text>
        </mj-column>
      </mj-section>
    </mj-container>
  </mj-body>
</mjml>
```

This entry point:

`index.js`
```javascript
import template from './template.mjml';
console.log(template);
```

The console output will be:

```html
<!doctype html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<title></title>
<style type="text/css">
  #outlook a { padding: 0; }
  .ReadMsgBody { width: 100%; }
  .ExternalClass { width: 100%; }
  .ExternalClass * { line-height:100%; }
  body { margin: 0; padding: 0; -webkit-text-size-adjust: 100%; -ms-text-size-adjust: 100%; }
  table, td { border-collapse:collapse; mso-table-lspace: 0pt; mso-table-rspace: 0pt; }
  img { border: 0; height: auto; line-height: 100%; outline: none; text-decoration: none; -ms-interpolation-mode: bicubic; }
  p { display: block; margin: 13px 0; }
</style>
<!--[if !mso]><!-->
<style type="text/css">
  @import url(https://fonts.googleapis.com/css?family=Ubuntu:300,400,500,700);
</style>
<style type="text/css">
  @media only screen and (max-width:480px) {
    @-ms-viewport { width:320px; }
    @viewport { width:320px; }
  }
</style>
<link href="https://fonts.googleapis.com/css?family=Ubuntu:300,400,500,700" rel="stylesheet" type="text/css">
<!--<![endif]-->
<style type="text/css">
  @media only screen and (min-width:480px) {
    .mj-column-per-100, * [aria-labelledby="mj-column-per-100"] { width:100%!important; }
  }
</style>
</head>
<body>
  <div><!--[if mso]>
      <table border="0" cellpadding="0" cellspacing="0" width="600" align="center" style="width:600px;"><tr><td>
      <![endif]--><div style="margin:0 auto;max-width:600px;"><table cellpadding="0" cellspacing="0" style="font-size:0;width:100%;" align="center" border="0"><tbody><tr><td style="text-align:center;vertical-align:top;font-size:0;padding:20px 0;"><!--[if mso]>
      <table border="0" cellpadding="0" cellspacing="0"><tr><td style="vertical-align:top;width:600px;">
      <![endif]--><div aria-labelledby="mj-column-per-100" class="mj-column-per-100" style="vertical-align:top;display:inline-block;font-size:13px;text-align:left;width:100%;"><table cellpadding="0" cellspacing="0" width="100%" border="0"><tbody><tr><td style="word-break:break-word;font-size:0;padding:10px 25px;" align="left"><div style="cursor:auto;color:#000000;font-family:Ubuntu, Helvetica, Arial, sans-serif;font-size:13px;line-height:22px;">Hello World!</div></td></tr></tbody></table></div><!--[if mso]>
      </td></tr></table>
      <![endif]--></td></tr></tbody></table></div><!--[if mso]>
      </td></tr></table>
      <![endif]-->
      <!--[if mso]>
      <table border="0" cellpadding="0" cellspacing="0" width="600" align="center" style="width:600px;"><tr><td>
      <![endif]--><div></div><!--[if mso]>
      </td></tr></table>
      <![endif]--></div>
</body>
</html>
```

<h2 align="center">Contribute</h2>

Feel free to dive in! [Open an issue](https://github.com/ramasilveyra/mjml-loader/issues/new) or submit PRs.

mjml-loader follows the [Contributor Covenant](http://contributor-covenant.org/version/1/3/0/) Code of Conduct.

<h2 align="center">License</h2>

[MIT © Ramiro Silveyra d'Avila](LICENSE.md)


CHANGELOG.md
<a name="1.0.1"></a>
## [1.0.1](https://github.com/ramasilveyra/mjml-loader/compare/v1.0.0...v1.0.1) (2017-10-03)


### Bug Fixes

* **package:** add build files to pkg ([12edd35](https://github.com/ramasilveyra/mjml-loader/commit/12edd35)), closes [#7](https://github.com/ramasilveyra/mjml-loader/issues/7)



<a name="1.0.0"></a>
# [1.0.0](https://github.com/ramasilveyra/mjml-loader/compare/v1.0.0-beta.1...v1.0.0) (2017-09-10)


### Features

* es modules support, drop webpack 1 support and build/tests/docs improvements ([de4b22d](https://github.com/ramasilveyra/mjml-loader/commit/de4b22d))

### Documentation

* Update README to use Webpack 2+ syntax ([6d62b72](https://github.com/ramasilveyra/mjml-loader/commit/6d62b72))



<a name="1.0.0-beta.1"></a>
# 1.0.0-beta.1 (2017-05-07)

* Changes based on code review. ([4a3f4ad](https://github.com/ramasilveyra/mjml-loader/commit/4a3f4ad))
* Don't use arrow function with the exports function as it does not get scoped to the webpack object,  ([3a0ca37](https://github.com/ramasilveyra/mjml-loader/commit/3a0ca37))
* Release 1.0.0-beta.1 ([86d4a72](https://github.com/ramasilveyra/mjml-loader/commit/86d4a72))
* Return html instead of mjml object to match loader's previous behaviour. ([4052851](https://github.com/ramasilveyra/mjml-loader/commit/4052851))
* Update mjml to 3.3.2. ([f93c21b](https://github.com/ramasilveyra/mjml-loader/commit/f93c21b))
* Update readme with correct loader reference and template is not a function. ([687bd41](https://github.com/ramasilveyra/mjml-loader/commit/687bd41))



<a name="1.0.0-beta.0"></a>
# 1.0.0-beta.0 (2016-07-23)

* Initial commit ([6b1d433](https://github.com/ramasilveyra/mjml-loader/commit/6b1d433))


test/index.test.js
import fs from 'fs';
import path from 'path';
import mjmlLoader from '../src';

const validMJMLPath = path.resolve(__dirname, 'fixtures/valid.mjml');
const validMJML = fs.readFileSync(validMJMLPath, 'utf-8');
const invalidMJMLPath = path.resolve(__dirname, 'fixtures/invalid.mjml');
const invalidMJML = fs.readFileSync(invalidMJMLPath, 'utf-8');

describe('while using mjml-loader', () => {
  it('should convert valid MJML to CommonJS module exporting the HTML result', () => {
    const context = {
      cacheable: jest.fn()
    };

    const result = mjmlLoader.call(context, validMJML);

    expect(result).toMatchSnapshot();
    expect(context.cacheable).toHaveBeenCalled();
  });

  it('should throw with invalid MJML', () => {
    const context = {
      cacheable: jest.fn(),
      resourcePath: invalidMJMLPath
    };

    const result = mjmlLoader.bind(context, invalidMJML);

    expect(result).toThrowError(/\[mjml-loader\] ERROR in /);
    expect(context.cacheable).toHaveBeenCalled();
  });
});


test/fixtures/entry.js
import mjmlFile from './valid.mjml';

export default mjmlFile;


test/functional.test.js
import path from 'path';
import del from 'del';
import webpack from 'webpack';

// TODO: For some reason jest in watch mode loops infinitely in this test.
// Needs investigation, skiped by default.
describe.skip('while using mjml-loader with raw-loader', () => {
  beforeEach(() => del(path.resolve(__dirname, 'fixtures/bundle.js')));

  it('should inline MJML result into the bundle', async () => {
    const options = {
      entry: path.resolve(__dirname, 'fixtures/entry.js'),
      output: {
        path: path.resolve(__dirname, 'fixtures'),
        filename: 'bundle.js',
        libraryTarget: 'commonjs2'
      },
      module: {
        rules: [{ test: /\.mjml$/, loader: require.resolve('../') }]
      },
      target: 'node'
    };

    await compile(options);
    const html = (await import('./fixtures/bundle')).default;

    expect(html).toMatchSnapshot();
  });
});

function compile(options) {
  return new Promise((resolve, reject) => {
    const compiler = webpack(options);
    compiler.run((err, stats) => {
      if (err || stats.hasErrors()) {
        return reject(err || stats);
      }

      return resolve(stats);
    });
  });
}


example/bundle.js
/******/ (function(modules) {
  // webpackBootstrap
  /******/ // The module cache
  /******/ var installedModules = {}; // The require function
  /******/
  /******/ /******/ function __webpack_require__(moduleId) {
    /******/
    /******/ // Check if module is in cache
    /******/ if (installedModules[moduleId]) {
      /******/ return installedModules[moduleId].exports;
      /******/
    } // Create a new module (and put it into the cache)
    /******/ /******/ var module = (installedModules[moduleId] = {
      /******/ i: moduleId,
      /******/ l: false,
      /******/ exports: {}
      /******/
    }); // Execute the module function
    /******/
    /******/ /******/ modules[moduleId].call(
      module.exports,
      module,
      module.exports,
      __webpack_require__
    ); // Flag the module as loaded
    /******/
    /******/ /******/ module.l = true; // Return the exports of the module
    /******/
    /******/ /******/ return module.exports;
    /******/
  } // expose the modules object (__webpack_modules__)
  /******/
  /******/
  /******/ /******/ __webpack_require__.m = modules; // expose the module cache
  /******/
  /******/ /******/ __webpack_require__.c = installedModules; // define getter function for harmony exports
  /******/
  /******/ /******/ __webpack_require__.d = function(exports, name, getter) {
    /******/ if (!__webpack_require__.o(exports, name)) {
      /******/ Object.defineProperty(exports, name, {
        /******/ configurable: false,
        /******/ enumerable: true,
        /******/ get: getter
        /******/
      });
      /******/
    }
    /******/
  }; // getDefaultExport function for compatibility with non-harmony modules
  /******/
  /******/ /******/ __webpack_require__.n = function(module) {
    /******/ var getter =
      module && module.__esModule
        ? /******/ function getDefault() {
            return module['default'];
          }
        : /******/ function getModuleExports() {
            return module;
          };
    /******/ __webpack_require__.d(getter, 'a', getter);
    /******/ return getter;
    /******/
  }; // Object.prototype.hasOwnProperty.call
  /******/
  /******/ /******/ __webpack_require__.o = function(object, property) {
    return Object.prototype.hasOwnProperty.call(object, property);
  }; // __webpack_public_path__
  /******/
  /******/ /******/ __webpack_require__.p = ''; // Load entry module and return exports
  /******/
  /******/ /******/ return __webpack_require__((__webpack_require__.s = 0));
  /******/
})(
  /************************************************************************/
  /******/ [
    /* 0 */
    /***/ function(module, __webpack_exports__, __webpack_require__) {
      'use strict';
      Object.defineProperty(__webpack_exports__, '__esModule', { value: true });
      /* harmony import */ var __WEBPACK_IMPORTED_MODULE_0__valid_mjml__ = __webpack_require__(1);

      console.log('mjmlFile', __WEBPACK_IMPORTED_MODULE_0__valid_mjml__['a' /* default */]);

      /***/
    },
    /* 1 */
    /***/ function(module, __webpack_exports__, __webpack_require__) {
      'use strict';
      /* harmony default export */ __webpack_exports__['a'] =
        '<!doctype html>\n<html xmlns="http://www.w3.org/1999/xhtml" xmlns:v="urn:schemas-microsoft-com:vml" xmlns:o="urn:schemas-microsoft-com:office:office">\n<head>\n  <title></title>\n  <!--[if !mso]><!-- -->\n  <meta http-equiv="X-UA-Compatible" content="IE=edge">\n  <!--<![endif]-->\n<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">\n<meta name="viewport" content="width=device-width, initial-scale=1.0">\n<style type="text/css">\n  #outlook a { padding: 0; }\n  .ReadMsgBody { width: 100%; }\n  .ExternalClass { width: 100%; }\n  .ExternalClass * { line-height:100%; }\n  body { margin: 0; padding: 0; -webkit-text-size-adjust: 100%; -ms-text-size-adjust: 100%; }\n  table, td { border-collapse:collapse; mso-table-lspace: 0pt; mso-table-rspace: 0pt; }\n  img { border: 0; height: auto; line-height: 100%; outline: none; text-decoration: none; -ms-interpolation-mode: bicubic; }\n  p { display: block; margin: 13px 0; }\n</style>\n<!--[if !mso]><!-->\n<style type="text/css">\n  @media only screen and (max-width:480px) {\n    @-ms-viewport { width:320px; }\n    @viewport { width:320px; }\n  }\n</style>\n<!--<![endif]-->\n<!--[if mso]>\n<xml>\n  <o:OfficeDocumentSettings>\n    <o:AllowPNG/>\n    <o:PixelsPerInch>96</o:PixelsPerInch>\n  </o:OfficeDocumentSettings>\n</xml>\n<![endif]-->\n<!--[if lte mso 11]>\n<style type="text/css">\n  .outlook-group-fix {\n    width:100% !important;\n  }\n</style>\n<![endif]-->\n\n<!--[if !mso]><!-->\n    <link href="https://fonts.googleapis.com/css?family=Ubuntu:300,400,500,700" rel="stylesheet" type="text/css">\n    <style type="text/css">\n\n        @import url(https://fonts.googleapis.com/css?family=Ubuntu:300,400,500,700);\n\n    </style>\n  <!--<![endif]--></head>\n<body>\n  \n  <div style="cursor:auto;color:#000000;font-family:Ubuntu, Helvetica, Arial, sans-serif;font-size:13px;line-height:22px;text-align:left;">Hello World!</div>\n</body>\n</html>';

      /***/
    }
    /******/
  ]
);
